### PR TITLE
Ab#10447 update advanced search

### PR DESCRIPTION
**Pull request by sharpedavid**
### Changes being made

Please include a short summary of the change.

### Context

What is the context for those changes?

### Quality Check

- [ ] E2E/ Unit/ Integration tests have been added.
- [ ] Frontend tests have been successfully run.
- [ ] Backend tests have been successfully run.
- [ ] The actual changes are separated from formatting changes.

### CODE REVIEW

File: backend/src/test/java/ca/bc/gov/hlth/mohums/user/UserServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+        int firstNameInvocations = firstName.isPresent() || search.isPresent() ? 1 : 0;
+        int lastNameInvocations = lastName.isPresent() || search.isPresent() ? 1 : 0;
+        int usernameInvocations = username.isPresent() || search.isPresent() ? 1 : 0;
+
+        Mockito.verify(userSpecifications, Mockito.times(1)).notServiceAccount();
+        Mockito.verify(userSpecifications, Mockito.times(1)).fromMohApplicationsRealm();
+        Mockito.verify(userSpecifications, Mockito.times(emailInvocations)).emailLike(ArgumentMatchers.anyString());
+        Mockito.verify(userSpecifications, Mockito.times(firstNameInvocations)).firstNameLike(ArgumentMatchers.anyString());
+        Mockito.verify(userSpecifications, Mockito.times(lastNameInvocations)).lastNameLike(ArgumentMatchers.anyString());
+        Mockito.verify(userSpecifications, Mockito.times(usernameInvocations)).usernameLike(ArgumentMatchers.anyString());
```
Comment (sharpedavid) [inline backend/src/test/java/ca/bc/gov/hlth/mohums/user/UserServiceTest.java:98]: Blackbox tests are best. They focus on what the app does, not how. Tests that check methods break easily and are hard to follow because they don't reflect real requirements.

If a test fails because I changed how something works, but not what it does, I'll drop the test. Ideally, our main functionality tests still cover us. This keeps testing straightforward and focused.

So I'm not saying to delete these tests, just to ensure that all the functional requirements are being tested. 

This kind of test, in my opinion, is only useful during development by the developer writing the code.

**Review by sharpedavid — COMMENTED** [review]
Everything looks like it's coming along very nicely. I'd like to go over the contents verbally so I can ask questions a bit more easily and hear the answers. So far everything is looking straight forward and well tests though, which is exactly how code should look.

For my review, I went through the files _once_ from top to bottom. My apologies if not all the comments or questions are clear. Let's discuss.

File: backend/src/main/java/ca/bc/gov/hlth/mohums/exceptions/RestExceptionHandler.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
+package ca.bc.gov.hlth.mohums.exceptions;
+
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;
+import org.springframework.http.ResponseEntity;
+
+@ControllerAdvice
+public class RestExceptionHandler extends ResponseEntityExceptionHandler {
```
[the remark this answers is not in the document]
Comment (filipflorek) [reply backend/src/main/java/ca/bc/gov/hlth/mohums/exceptions/RestExceptionHandler.java:10]: Good point. I'll separate it. I wanted to include exception handling so that the same error response as in Keycloak API is returned when user is not found. But this error handling only applies to GET `"/user/{userId}"` - I started with this simple endpoint to test out JPA integration, instead of jumping to the more complex GET `"/users"` endpoint right away.

File: backend/src/main/java/ca/bc/gov/hlth/mohums/user/UserEntity.java [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+    protected boolean enabled;
+    @Column(name = "EMAIL_VERIFIED")
+    protected boolean emailVerified;
+    @Column(name = "REALM_ID")
+    protected String realmId;
+
+    @OneToMany(cascade = CascadeType.REMOVE, orphanRemoval = true, mappedBy = "user")
+    @Fetch(FetchMode.SELECT)
+    @BatchSize(size = 20)
+    protected Collection<UserAttributeEntity> attributes = new LinkedList<>();
```
[the remark this answers is not in the document]
Comment (filipflorek) [reply backend/src/main/java/ca/bc/gov/hlth/mohums/user/UserEntity.java:39]: I copied it from the source code and will clean this up.

File: backend/src/main/java/ca/bc/gov/hlth/mohums/user/UserAttributeEntity.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
+package ca.bc.gov.hlth.mohums.user;
+
+import javax.persistence.*;
+
+@Entity
+@Table(name = "USER_ATTRIBUTE", schema = "KEYCLOAK")
+public class UserAttributeEntity {
```
[the remark this answers is not in the document]
Comment (filipflorek) [reply backend/src/main/java/ca/bc/gov/hlth/mohums/user/UserAttributeEntity.java:7]: Correct. I did not check if it's available as a library, but I'm making small changes to those Entity classes (For example bidirectional relationship between UserEntity and UserRoleMapping - so that URM can be referenced from UserEntity when preparing JOIN statement.

File: backend/src/main/java/ca/bc/gov/hlth/mohums/user/UserService.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+            }
+            if (username.isPresent()) {
+                userSpec.and(userSpecifications.usernameLike(username.get()));
+            }
+            if (email.isPresent()) {
+                userSpec.and(userSpecifications.emailLike(email.get()));
+            }
+        }
+
+        //TODO: should the block below be nested - there's possibility of manually calling API with search + organization, but not through UMC - does anyone else uses it?
```
[the remark this answers is not in the document]
Comment (filipflorek) [reply backend/src/main/java/ca/bc/gov/hlth/mohums/user/UserService.java:57]: This should be easy to accommodate, I guess the same applies for `search` + roles/login dates. But if I have more trouble than expected, UnsupportedOperationException is a great suggestion.
I will also reach out to PIDP/HCAP to get a better understanding on how they're using it (if at all).

File: backend/src/main/java/ca/bc/gov/hlth/mohums/user/UserService.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+        if(organizationId.isPresent()) {
+            userSpec.and(userSpecifications.hasOrganizationWithGivenId(organizationId.get()));
+        }
+        return userRepository.findAll(userSpec).stream().map(this::convertToDTO).collect(Collectors.toList());
+    }
+
+    /*
+    Convert UserEntity to match response structure returned by Keycloak API
```
[the remark this answers is not in the document]
Comment (filipflorek) [reply backend/src/main/java/ca/bc/gov/hlth/mohums/user/UserService.java:72]: Right, I wanted to keep the structure of response intact. The DTO will also include `lastLogDate` and `roles` fields in some search cases

File: backend/src/main/java/ca/bc/gov/hlth/mohums/user/UserService.java [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+public class UserService {
+
+    private final UserRepository userRepository;
+    private final UserSpecifications userSpecifications;
```
[the remark this answers is not in the document]
Comment (filipflorek) [reply backend/src/main/java/ca/bc/gov/hlth/mohums/user/UserService.java:14]: Spring docs
https://docs.spring.io/spring-data/jpa/reference/jpa/specifications.html

I understand that Specifcations are a wrapper around Criteria API, that allow you to chain and reuse them, so this link has some more examples or raw Criteria usage.
https://www.baeldung.com/hibernate-criteria-queries

This had some good real-world examples with JOIN clauses etc.
https://medium.com/@bubu.tripathy/dynamic-query-with-specification-interface-in-spring-data-jpa-ae8764e32162

File: backend/src/test/java/ca/bc/gov/hlth/mohums/user/UserRepositoryTest.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+
+    @Autowired
+    private UserRepository userRepository;
+    @Autowired
+    private UserSpecifications userSpecifications;
+
+    @Test
+    public void testFindByIdSuccessful() {
+        String umsTestUserId = "c35d48ea-3df9-4758-a27b-94e4cab1ba44";
```
[the remark this answers is not in the document]
Comment (filipflorek) [reply backend/src/test/java/ca/bc/gov/hlth/mohums/user/UserRepositoryTest.java:34]: The user? It was manually created a while ago in dev env for the purpouse of UMS tests.

File: backend/src/test/java/ca/bc/gov/hlth/mohums/user/UserServiceTest.java (hunk 1 above)
Comment (filipflorek) [reply backend/src/test/java/ca/bc/gov/hlth/mohums/user/UserServiceTest.java:98]: Thanks, good point. I think I wanted to somehow test the method because it's an important one, but did not find a meaningful way to test it. Now I understand what you mean. Every time I add more params to the function or change the logic a bit I need to change the test as well  - so probably will delete it.

File: backend/src/test/java/ca/bc/gov/hlth/mohums/user/UserRepositoryTest.java [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    public void testFindAllByOrganizationNoResults() {
+        Specification<UserEntity> userSpec = baseUserSpecification()
+                .and(userSpecifications.hasOrganizationWithGivenId("non-existing-org-id"));
+        List<UserEntity> result = userRepository.findAll(userSpec);
+
+        assertThat(result.isEmpty()).isTrue();
+    }
+
+    @ParameterizedTest
+    @MethodSource("provideUsernameForFindAllByUsername")
```
[the remark this answers is not in the document]
Comment (filipflorek) [reply backend/src/test/java/ca/bc/gov/hlth/mohums/user/UserRepositoryTest.java:125]: https://www.baeldung.com/parameterized-tests-junit-5

If tests are simpler, you can provide values right at the annotation level. I like to use it to test many cases without much boilerplate code.

**Review by ahoplock — APPROVED** [review]
Reviewed and approved. Deployed through to Production.
